rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,382 @@
+$ ls -l /usr/local/bin/teleport-updater
+/usr/local/bin/teleport-updater -> /var/lib/teleport/versions/15.0.0/bin/teleport-updater
+$ ls -l /usr/local/lib/systemd/system/teleport.service
+/usr/local/lib/systemd/system/teleport.service -> /var/lib/teleport/versions/15.0.0/etc/systemd/teleport.service
+```
+
+updates.yaml:
+```
+version: v1
+kind: agent_versions
+spec:
+  # proxy specifies the Teleport proxy address to retrieve the agent version and update configuration from.
+  proxy: mytenant.teleport.sh
+  # enabled specifies whether auto-updates are enabled, i.e., whether teleport-updater update is allowed to update the agent.
+  enabled: true
+  # active_version specifies the active (symlinked) deployment of the telepport agent.
+  active_version: 15.1.1
+```
+
+backup.yaml:
+```
+version: v1
+kind: config_backup
+spec:
+  # proxy address from the backup
+  proxy: mytenant.teleport.sh
+  # version from the backup
+  version: 15.1.0
+  # time the backup was created
+  creation_time: 2020-12-09T16:09:53+00:00
+```
+
+### Runtime
+
+The agent-updater will run as a periodically executing systemd service which runs every 10 minutes.
+The systemd service will run:
+```shell
+$ teleport-updater update
+```
+
+After it is installed, the `update` subcommand will no-op when executed until configured with the `teleport-updater` command:
+```shell
+$ teleport-updater enable --proxy mytenant.teleport.sh
+```
+
+If the proxy address is not provided with `--proxy`, the current proxy address from `teleport.yaml` is used.
+
+On servers without Teleport installed already, the `enable` subcommand will change the behavior of `teleport-update update` to update teleport and restart the existing agent, if running.
+It will also run update teleport immediately, to ensure that subsequent executions succeed.
+
+The `enable` subcommand will:
+1. Configure `updates.yaml` with the current proxy address and set `enabled` to true.
+2. Query the `/v1/webapi/ping` endpoint.
+3. If the current updater-managed version of Teleport is the latest, and teleport package is not installed, quit.
+4. If the current updater-managed version of Teleport is the latest, but the teleport package is installed, jump to (14).
+5. Ensure there is enough free disk space to upgrade Teleport.
+6. Download the desired Teleport tarball specified by `agent_version` and `server_edition`.
+7. Download and verify the checksum (tarball URL suffixed with `.sha256`).
+8. Extract the tarball to `/var/lib/teleport/versions/VERSION`.
+9. Replace any existing binaries or symlinks with symlinks to the current version.
+10. Backup /var/lib/teleport into `/var/lib/teleport/versions/OLD-VERSION/backup/teleport`
+11. Restart the agent if the systemd service is already enabled.
+12. Set `active_version` in `updates.yaml` if successful or not enabled.
+13. Replace the symlinks/binaries and `/var/lib/teleport` and quit (exit 1) if unsuccessful.
+14. Remove any `teleport` package if installed.
```

Comment:
Package managers (at least APT/dpkg) keep track of which file was brought by which package and might not remove all files immediately (but will cleanup later). We must be careful about how `apt remove` and `apt purge` work, we got bit by this last time. `remove` might leave files, we reuse them, then the next purge/cleanup will trash important parts such as `teleport.yaml` or the teleport state and the agent will be bricked.

Reply:
Added